docs/proposed/conflict-api.rst

```diff
@@ -11,42 +11,77 @@ This document is aimed at programmers working on magic-folder. It is a proposed
 Motivation
 ==========
 
-It is desirable to have an explicit API for dealing with conflicts.
-This allows UI programs to be sure they're accurately indicating intent to magic-folders.
+Under normal operation, magic-folders scans other participants' Snapshots and reflects those changes locally; sometimes, these changes can conflict.
+
+It is desirable to have an explicit API for noticing and resolving conflicts.
+The core of magic-folders operation with users is the filesystem.
+This makes the filesystem also an API and conflicts must be part of that API.
+The command-line interface shall have sub-commands for listing and resolving conficts; these commands will use the HTTP APIs.
+
+
+Filesystem API for Conflicts and Resolution
+===========================================
+
+When a Snapshot is found to conflict with a particluar local file (say ``foo``) a "conflict file" is written beside it, reflecting the other participants' content (``foo.conflict-laptop``).
+That is, ``foo.conflict-laptop`` indicates that the participant "``laptop``" has a conflicting update
```

Comment:
Right now, these files are not treated specially by magic-folder, so are themselves snapshotted. That isn't entirely unreasonable (it is what [syncthing](https://docs.syncthing.net/users/syncing.html#conflict-handling) does, for example). However, if we do want that, we should be explicit about it.

If we do treat them normally, we likely want to explicitly record them when we create them (which means we could store the conflicted remote snapshot under the conflict name in our personal DMD). This also leads me to wonder if we should store the local version under a conflict name as well, when we trigger a conflict (so we'd end up with an identical set of `.conflict-*` files on every device, which we'd eventually get anyway, as we sync between devices).

Also, do we want to handle snapshotting a file specially (i.e. include some extra metadata), when we take a snapshot while conflict files exists?

If we *don't* want to sync `.conflict-*` files as normal files, we should file an issue so that we can implement that change.

Reply:
There is also the question if a particular particpant (say `laptop`) update there version of `foo` and and also `foo.conflict-laptop` is update differently? Even we we don't snapshot `.conflict-*` files, we need to address what happens if somebody makes local changes to one, and the corresponding remote update the file. Particularly the HTTP conflict resolution method seems to indicate that making edits to a conflict file and then calling that API is an expected way to deal with conflicts.

Reply:
> If we don't want to sync .conflict-* files as normal files, we should file an issue so that we can implement that change.

Yes, this is what we want to do. There's at least a comment to that effect in the uploader i believe.

The only file that should ever be edited is the base one (e.g. `foo`). If we want a thing besides "take all changes from X" (including "us") then that's a new HTTP API that should accept a body that is the new text (i.e. distinct from the file-system-based API where you'd edit `foo` to contain the desired resolution, then delete all the conflict files).

Reply:
> There is also the question if a particular particpant (say laptop) update there version of foo and and also foo.conflict-laptop is update differently?

Can't fully parse that sentence, but once there is a conflict, the downloader stops downloading things for that file so it shouldn't be "updating" any existing conflict-files (and you shouldn't edit them locally, but if you do nothing will happen).

Reply:
> > There is also the question if a particular particpant (say laptop) update there version of foo and and also foo.conflict-laptop is update differently?
> 
> Can't fully parse that sentence, but once there is a conflict, the downloader stops downloading things for that file so it shouldn't be "updating" any existing conflict-files 

As long as we record the appropriate parent when we resolve a conflict, we can probably get away with not updating the conflict file if the peer updates their version. That will mean the user will need to immediately resolve another conflict with the same peer, though.

If the behavior you've described is intended, it should be documented. I had assumed that it might make sense to update the `.conflict-*` file if the peer made more changes and the `.conflict-*` file was unchanged locally. If we could magically notice when a user realized there was conflict and started trying to resolve it, the ideal behavior would likely be to keep the `.conflict-*` up-to-date with the peer until the user started trying to resolve it. Given that magic is impossible, I'm not sure where on the scale of keeping the first version and keeping it up-to-date makes the most sense as a default.

> (and you shouldn't edit them locally, but if you do nothing will happen).

I've often found it useful to edit both files, when trying to merge changes, so I don't think it is reasonable to expect the user to not edit them locally.

> once there is a conflict, the downloader stops downloading things for that file

The implementation in #540 seems to force conflicts to always be resolved pair-wise. I'd've assumed that you'd want to at least publish the conflicts with every peer we can find, rather than just the first one we encounter, even if we don't update them in response to further changes in the peer.

Reply:
There is no way to resolve conflicts implemented in #540

Reply:
> I've often found it useful to edit both files, when trying to merge changes, so I don't think it is reasonable to expect the user to not edit them locally.

Obviously you can do whatever you like, but there's no semantics around doing so (is what I'm saying above).